src/cascadia/PublicTerminalCore/Source.cpp

```diff
@@ -0,0 +1,291 @@
+    wc.lpszClassName = szTOOLBOXWNDCLASS;
+
+    bool registered = RegisterClassW(&wc) != 0;
+
+    if (registered)
+    {
+        auto h = CreateWindowExW(
+            0,
+            szTOOLBOXWNDCLASS,
+            nullptr,
+            WS_CHILD |
+                WS_CLIPCHILDREN |
+                WS_CLIPSIBLINGS |
+                WS_VISIBLE,
+            0,
+            0,
+            50,
+            50,
+            parentHwnd,
+            nullptr,
+            GetModuleHandle(nullptr),
+            nullptr);
+
+        _hwnd = h;
+
+        SetWindowLongPtr(_hwnd, GWLP_USERDATA, (LONG_PTR)this);
+    }
+}
+
+HwndTerminal::~HwndTerminal()
+{
+    DestroyWindow(_hwnd);
+}
+
+HRESULT HwndTerminal::Initialize()
+{
+    _terminal = std::make_unique<::Microsoft::Terminal::Core::Terminal>();
+    auto renderThread = std::make_unique<::Microsoft::Console::Render::RenderThread>();
+    auto* const localPointerToThread = renderThread.get();
+    _renderer = std::make_unique<::Microsoft::Console::Render::Renderer>(_terminal.get(), nullptr, 0, std::move(renderThread));
+
+    RETURN_IF_FAILED(localPointerToThread->Initialize(_renderer.get()));
+
+    auto dxEngine = std::make_unique<::Microsoft::Console::Render::DxEngine>();
+    RETURN_IF_FAILED(dxEngine->SetHwnd(_hwnd));
+    RETURN_IF_FAILED(dxEngine->Enable());
+    _renderer->AddRenderEngine(dxEngine.get());
+
+    auto pfn = std::bind(&::Microsoft::Console::Render::Renderer::IsGlyphWideByFont, _renderer.get(), std::placeholders::_1);
+    SetGlyphWidthFallback(pfn);
+
+    _UpdateFont(96);
+    RECT windowRect;
+    GetWindowRect(_hwnd, &windowRect);
+
+    const COORD windowSize{ static_cast<short>(windowRect.right - windowRect.left), static_cast<short>(windowRect.bottom - windowRect.top) };
+
+    // Fist set up the dx engine with the window size in pixels.
+    // Then, using the font, get the number of characters that can fit.
+    // Resize our terminal connection to match that size, and initialize the terminal with that size.
+    const auto viewInPixels = Viewport::FromDimensions({ 0, 0 }, windowSize);
+    RETURN_IF_FAILED(dxEngine->SetWindowSize({ viewInPixels.Width(), viewInPixels.Height() }));
+
+    _renderEngine = std::move(dxEngine);
+
+    _terminal->SetBackgroundCallback([](auto) {});
+
+    _terminal->Create(COORD{ 80, 25 }, 1000, *_renderer);
+    _terminal->SetDefaultBackground(RGB(5, 27, 80));
+    _terminal->SetDefaultForeground(RGB(255, 255, 255));
+
+    localPointerToThread->EnablePainting();
+
+    return S_OK;
+}
+
+void HwndTerminal::RegisterScrollCallback(std::function<void(int, int, int)> callback)
+{
+    _terminal->SetScrollPositionChangedCallback(callback);
+}
+
+void HwndTerminal::_UpdateFont(int newDpi)
+{
+    auto lock = _terminal->LockForWriting();
+
+    // TODO: MSFT:20895307 If the font doesn't exist, this doesn't
+    //      actually fail. We need a way to gracefully fallback.
+    _renderer->TriggerFontChange(newDpi, _desiredFont, _actualFont);

```

Comment:
I don't like this copy paste. Where is it coming from and why can't this just be defined by inheritance from that?

Reply:
This is coming from `TerminalControl.cpp`. We can't reuse it via inheritance since `TerminalControl` is tuned specifically for the UWP control.